File: uapc/archive/2022/problems/numbercolosseum/submissions/accepted/numbercolosseum-jc.py
#! /usr/bin/python3

L = []

def same_sign(a, b):
    return (a < 0) == (b < 0)

def push(x):
    while x:
        if not L or same_sign(x, L[-1]):
            L.append(x)
            break
        x += L.pop()

n = int(input())
for x in map(int, input().split()):
    push(x)

if not L:
    print("Tie!")
else:
    print("Positives" if L[0] > 0 else "Negatives", "win!")
    print(' '.join(map(str, L)))

File: uapc/archive/2022/problems/numbercolosseum/submissions/accepted/numbercolosseum-jc.cpp
#include <iostream>
#include <deque>
#include <algorithm>
#include <iterator>

using namespace std;

#define same_sign(a, b) ((a<0) == (b<0))

deque<int> Q;

void push(int x) {
    if (x == 0) return;

    if (Q.empty() || same_sign(x, Q.back())) {
        Q.push_back(x);
    } else {
        x += Q.back();
        Q.pop_back();
        push(x);
    }
}

int main() {
    int n, x;
    cin >> n;
    while (n--) {
        cin >> x;
        push(x);
    }
    if (Q.empty()) {
        cout << "Tie!" << endl;
    } else {
        cout << (Q.front() > 0 ? "Positives" : "Negatives") << " win!" << endl;
        cout << Q[0];
        for (int i = 1; i < Q.size(); ++i)
            cout << ' ' << Q[i];
        cout << endl;
    }
    return 0;
}
